### PR TITLE
Daniel convergence29ã032018

### PR DESCRIPTION
**Pull request by DanielBiman**
This function is able to test the convergence of classifikation and regression tasks. Should we add bbmisc and plyr to the description of our package? Next step is to add the cluster test. After that we can think of a smooth way to visualice the result of cluster/classification.

### CODE REVIEW

File: R/benchmarks.R [hunk 1]
```diff
@@ -2,23 +2,52 @@
...
+      geom_point(aes(x = seq_along(values), y = values, colour = "Sum of Shapley values")) +
+      geom_line(aes(x = seq_along(values), y = cumsum(values)/seq_along(values), colour = "Moving Averages")) +
+      geom_line(aes(x = seq(1:convergence.iterations), y = rep(response, convergence.iterations), colour = "Prediction")) +
+      scale_colour_discrete(name = NULL) + labs(x = "Convergence iterations", y = "Shapley value") +
+      theme(legend.position="bottom")
+    return(plot)
+  }
+  if(return.value == "values") {
+    if(getTaskType(task) == "classif"){
+      result = plyr::count(as.data.frame(values))
```
Comment (redichh) [inline R/benchmarks.R]: Can you solve it without an extra package? According to our previous lessons we should avoid dependiencies as far as it is possible. The following (not complete) code shows, that we can get fine with the basic methods:
`> a = as.data.frame(rbind("A", "A", "B", "D", "A", "D"))`
`> names(a) = "values"`
`> summary(a$values)`
`A B D `
`3 1 2 `
And with `summary(a$values)["A"]` you can get `A 3` as a result. So I think this should be the same as plyr::count but needs a little bit more code

File: R/benchmarks.R [hunk 2]
```diff
@@ -2,23 +2,52 @@
...
-#a.cum = cumsum(a) / seq_along(a)
-#a.sub = a
-#ggplot() +
-#  geom_point(aes(x = seq_along(a), y = a)) +
-#  geom_line(aes(x = seq_along(a), y = cumsum(a)/seq_along(a)), color = "red")
+  values = rep(0, times = convergence.iterations)
+  for(i in 1:convergence.iterations) {
+    if(getTaskType(task) == "classif"){
+      shap = shapley(row.nr, task, model, iterations)
+      values[i] = shap$"_Class"[BBmisc::getMaxIndex(rowSums(shap[,getTaskFeatureNames(task)]))]
```
Comment (redichh) [inline R/benchmarks.R]: Same as below: please try to find a way without other packages. We should avoid dependencies

File: R/benchmarks.R [hunk 3]
```diff
@@ -2,23 +2,52 @@
...
-#a = test.covergence(196, iterations = 20)
-#a.cum = cumsum(a) / seq_along(a)
-#a.sub = a
-#ggplot() +
-#  geom_point(aes(x = seq_along(a), y = a)) +
-#  geom_line(aes(x = seq_along(a), y = cumsum(a)/seq_along(a)), color = "red")
+  values = rep(0, times = convergence.iterations)
+  for(i in 1:convergence.iterations) {
+    if(getTaskType(task) == "classif"){
+      shap = shapley(row.nr, task, model, iterations)
```
Comment (redichh) [inline R/benchmarks.R]: This is the same as line 28. It doesn't matter what condition we have, we always do this. So can you put that line before the `if(...)`-block into line 23 but after the `for`-loop

File: R/benchmarks.R [hunk 4]
```diff
@@ -2,23 +2,52 @@
...
 
-  return(values)
-}
+  assert_number(iterations, convergence.iterations, row.nr)
+  assert_choice(return.value, c("plot", "values"))
+  assert_class(model, "WrappedModel")
+
+  data = getTaskData(task)[row.nr,]
+  prediction = getPredictionData(data, model, task)
+  prediction.class = names(prediction[BBmisc::getMaxIndex(prediction)])
```
Comment (redichh) [inline R/benchmarks.R]: Same here as below. Please find a way without a further package

**Review by redichh — CHANGES_REQUESTED** [review]

File: R/benchmarks.R (hunk 3 above)
Comment (DanielBiman) [reply R/benchmarks.R]: Alright.

File: R/benchmarks.R (hunk 1 above)
Comment (DanielBiman) [reply R/benchmarks.R]: Done.

File: R/benchmarks.R (hunk 2 above)
Comment (DanielBiman) [reply R/benchmarks.R]: Done.

File: R/benchmarks.R (hunk 4 above)
Comment (DanielBiman) [reply R/benchmarks.R]: Done.

File: R/benchmarks.R [hunk 5]
```diff
@@ -41,9 +42,10 @@ test.convergence = function(row.nr=2, convergence.iterations = 20, iterations =
   }
   if(return.value == "values") {
     if(getTaskType(task) == "classif"){
-      result = plyr::count(as.data.frame(values))
-      result$values = as.character(result$values)
-      result$values[match(prediction.class, result$values)] = paste0(">>", prediction.class, "<<")
+      values = replace(values, grep(prediction.class, values), paste0(">>", prediction.class, "<<"))
+      result = as.data.frame(values)
+      names(result) = "values"
```
Comment (redichh) [inline R/benchmarks.R]: Ich habe herausgefunden, dass man bei `as.data.frame` den parameter `row.names` übergeben kann. Damit können wir uns diese Zeile hier sparen. Und an mindestens 4 Stellen in meinem Code auch :)

**Pull request merged**